google/cloud/bigtable/_row_merger.py

```diff
@@ -0,0 +1,621 @@
+    AsyncGenerator,
+)
+
+"""
+This module provides a set of classes for merging ReadRowsResponse chunks
+into Row objects.
+
+- RowMerger is the highest level class, providing an interface for asynchronous
+  merging with or without retrues
+- StateMachine is used internally to track the state of the merge, including
+  rows the current row and the keys of the rows that have been processed.
+  It processes a stream of chunks, and will raise InvalidChunk if it reaches
+  an invalid state.
+- State classes track the current state of the StateMachine, and define what
+  to do on the next chunk.
+- RowBuilder is used by the StateMachine to build a Row object.
+"""
+
+
+class _RowMerger(AsyncIterable[Row]):
+    """
+    RowMerger handles the logic of merging chunks from a ReadRowsResponse stream
+    into a stream of Row objects.
+
+    RowMerger.merge_row_response_stream takes in a stream of ReadRowsResponse
+    and handles turns them into a stream of Row objects using an internal
+    StateMachine.
+
+    RowMerger(request, client) handles row merging logic end-to-end, including
+    performing retries on stream errors.
+    """
+
+    def __init__(
+        self,
+        request: dict[str, Any],
+        client: BigtableAsyncClient,
+        *,
+        cache_size: int = 0,
+        operation_timeout: float | None = None,
+        per_row_timeout: float | None = None,
+        per_request_timeout: float | None = None,
+        revise_on_retry: bool = True,
+    ):
+        """
+        Args:
+          - request: the request dict to send to the Bigtable API
+          - client: the Bigtable client to use to make the request
+          - cache_size: the size of the buffer to use for caching rows from the network
+          - operation_timeout: the timeout to use for the entire operation, in seconds
+          - per_row_timeout: the timeout to use when waiting for each individual row, in seconds
+          - revise_on_retry: if True, retried request will be modified based on rows that have already been seen
+        """
+        self.last_seen_row_key: bytes | None = None
+        self.emitted_rows: Set[bytes] = set()
+        cache_size = max(cache_size, 0)
+        self.request = request
+        self.operation_timeout = operation_timeout
+        # lock in paramters for retryable wrapper
+        self.partial_retryable = partial(
+            self.retryable_merge_rows,
+            client.read_rows,
+            cache_size,
+            per_row_timeout,
+            per_request_timeout,
+            revise_on_retry,
+        )
+        predicate = retries.if_exception_type(
+            InvalidChunk,
+            core_exceptions.ServerError,
+            core_exceptions.TooManyRequests,
+        )
+
+        def on_error_fn(exc):
+            if predicate(exc):
+                self.errors.append(exc)
+
+        retry = retries.AsyncRetry(
+            predicate=predicate,
+            timeout=self.operation_timeout,
+            initial=0.1,
```

Comment:
Is the time unit here also seconds? maximum of 1 second seems too short. should we update it to 1 minute?

Reply:
Good catch, I was meaning to update these to Java's values before merging.

Does initial=0.1, multiplier=2, maximum=60 sound better?

Reply:
can we make initial 0.01 (10 ms)? rest sounds good. Also, are we adding jitter ?

Reply:
Yes, retries use the `api-core` standard `exponential_sleep_generator` internally, and it looks like it [returns a random value between 0 and the next value](https://github.com/googleapis/python-api-core/blob/5dfb6afda81bcc13e3432b020e144f7873fe1f4e/google/api_core/retry.py#L144) at each sleep step.

We could easily swap in sleep logic if needed too, but this seems to be the standard used by other libraries